### PR TITLE
Add recurrent gemma

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Adds support for recurrent gemma

### CODE REVIEW

File: src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            This number must divide `hidden_size` and `lru_width`.
+        lru_width (`int` or `None`, *optional*):
+            Dimension of the hidden representations of the RG-LRU. If `None`
+            this will be set to `hidden_size`.
+            Whether to scale the output of the embeddings by `sqrt(hidden_size)`.
+        embeddings_scale_by_sqrt_dim (`<fill_type>`, *optional*, defaults to `True`): <fill_docstring>
+        attention_window_size (`int`, *optional*, defaults to 2048):
+            The size of the attention window used in the attention block.
+        conv1d_width (`<fill_type>`, *optional*, defaults to 4): <fill_docstring>
+        logits_soft_cap (`<fill_type>`, *optional*, defaults to 30.0): <fill_docstring>
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py:58]: Make sure to run make fixup after merging this

```suggestion
        logits_soft_cap (`float`, *optional*, defaults to 30.0):
            The value at which the logits should be soft-capped to after the transformer and LM-head computation in the Causal LM architecture.
```

File: src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            relevant if `config.is_decoder=True`.
+        pad_token_id (`int`, *optional*, defaults to 0):
+            Padding token id.
+        eos_token_id (`int`, *optional*, defaults to 1):
+            End of stream token id.
+        bos_token_id (`int`, *optional*, defaults to 2):
+            Beginning of stream token id.
+        tie_word_embeddings (`bool`, *optional*, defaults to `True`):
+            Whether to tie weight embeddings
+        hidden_activation (`<fill_type>`, *optional*, defaults to `"gelu_pytorch_tanh"`): <fill_docstring>
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py:73]:
```suggestion
        hidden_activation (``str` or `function``, *optional*, defaults to `"gelu_pytorch_tanh"`):
            The hidden activation used in the recurrent block as well as the MLP layer of the decoder layers.
```

File: src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        partial_rotary_factor (`<fill_type>`, *optional*, defaults to 0.5): <fill_docstring>
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py:74]:
```suggestion
        partial_rotary_factor (`float`, *optional*, defaults to 0.5):
            The partial rotary factor used in the initialization of the rotary embeddings.
```

File: src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        rope_theta (`float`, *optional*, defaults to 10000.0):
+            The base period of the RoPE embeddings.
+        block_types (`<fill_type>`, *optional*, defaults to `('recurrent', 'recurrent', 'attention')`): <fill_docstring>
+        attention_dropout (`<fill_type>`, *optional*, defaults to 0.0): <fill_docstring>
+        num_key_value_heads (`<fill_type>`, *optional*, defaults to 16): <fill_docstring>
+        attention_bias (`<fill_type>`, *optional*, defaults to `False`): <fill_docstring>
+        w_init_variance_scale (`<fill_type>`, *optional*, defaults to 0.01): <fill_docstring>
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py:81]: To update

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+        return attn_output
+
+    def _setup_cache(self, batch_size, device, dtype=None):
+        self.dtype = dtype if dtype is not None else torch.float32
+        cache_shape = (batch_size, self.num_key_value_heads, self.config.attention_window_size, self.head_dim)
+        self.value_states = torch.zeros(cache_shape, dtype=self.dtype, device=device)
+        self.key_states = torch.zeros(cache_shape, dtype=self.dtype, device=device)
+
+    @torch.no_grad()
+    def _update_cache(self, key_states, value_states, **cache_kwargs):
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:226]: Great!

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+    config_class = RecurrentGemmaConfig
+    base_model_prefix = "model"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["RecurrentGemmaDecoderLayer"]
+    _skip_keys_device_placement = ["cache"]
+    _supports_flash_attn_2 = False
+    _supports_sdpa = False  # we can't compare with eager for now
+    _supports_cache_class = True
+
+    def _init_weights(self, module):
```
Comment (LysandreJik) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:543]: Intense!

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+            choice_labels,
+        ) = config_and_inputs
+        inputs_dict = {"input_ids": input_ids, "attention_mask": input_mask}
+        return config, inputs_dict
+
+
+@require_torch
+class RecurrentGemmaModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin, unittest.TestCase):
+    all_model_classes = (RecurrentGemmaForCausalLM,) if is_torch_available() else ()
+    # all_generative_model_classes = (RecurrentGemmaForCausalLM,) if is_torch_available() else () #TODO @gante not fully supported
```
Comment (LysandreJik) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:288]: It doesn't work with generate yet?

**Review by LysandreJik — APPROVED** [review]
Ok great, LGTM!

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py (hunk 7 above)
Comment (ArthurZucker) [reply tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:288]: It does but a lot of tests are not working (20) because they rely on past key values

File: src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+    def __init__(
+        self,
+        num_hidden_layers=26,
+        vocab_size=256000,
+        hidden_size=2560,
+        intermediate_size=3 * 2560,
+        num_attention_heads=10,
+        lru_width=None,
+        embeddings_scale_by_sqrt_dim=True,
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/configuration_recurrent_gemma.py:110]: No more you are right!

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+    def test_model_parallelism(self):
+        pass
+
+    def test_model_parallel_beam_search(self):
+        pass
+
+    def _check_past_key_values_for_generate(self, *args, **kwargs):
+        return True
+
+    @unittest.skip("Seems to rely on `past_key_values`")
```
Comment (amyeroberts) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:385]: "Seems to" is quite vague!

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+    def test_assisted_decoding_matches_greedy_search(self):
+        pass
+
+    @unittest.skip("recurrent gemma's output different if you pad lefr or right. This is expected")
```
Comment (amyeroberts) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:389]: nit - consistency & spelling

```suggestion
    @unittest.skip("RecurrentGemma's output different if you pad left or right. This is expected")
```

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+    def test_left_padding_compatibility(self):
+        pass
+
+    @unittest.skip("Seems to rely on `past_key_values`")
```
Comment (amyeroberts) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:393]: Same here - reason isn't clear and not strong enough for skipping a test

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 12]
```diff
@@ -0,0 +1,938 @@
...
+            scale=self.head_dim**-0.5,
+        )
+
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.view(bsz, q_len, self.hidden_size)
+        attn_output = self.o_proj(attn_output)
+        return attn_output
+
+    def _setup_cache(self, batch_size, device, dtype=None):
+        self.dtype = dtype if dtype is not None else torch.float32
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:220]: Do we want to default here to `torch.float32` or torch's default dtype?

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 13]
```diff
@@ -0,0 +1,938 @@
...
+        cache_shape = (batch_size, self.num_key_value_heads, self.config.attention_window_size, self.head_dim)
+        self.value_states = torch.zeros(cache_shape, dtype=self.dtype, device=device)
+        self.key_states = torch.zeros(cache_shape, dtype=self.dtype, device=device)
+
+    @torch.no_grad()
+    def _update_cache(self, key_states, value_states, **cache_kwargs):
+        """
+        torch.compile compatible sliding window.
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:228]: Nice :)

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 14]
```diff
@@ -0,0 +1,938 @@
...
+        """The backward pass, which clips the `sqrt` gradient."""
+        (x,) = ctx.saved_tensors
+        clipped_x_times_4 = torch.clip(4.0 * x, min=1 / (_MAX_SQRT_GRADIENT**2))
+        return grad_output / torch.sqrt(clipped_x_times_4)
+
+
+class RecurrentGemmaRglru(nn.Module):
+    """A Real-Gated Linear Recurrent Unit (RG-LRU) layer."""
+
+    recurrent_states = None
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:284]: Why as a class attribute instead of an instance attribute?

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 15]
```diff
@@ -0,0 +1,938 @@
...
+                return contextualized_states.type(hidden_states.dtype), contextualized_states[:, -1]
+
+        else:
+            # Using scan in linear mode.
+            if recurrent_states is None:
+                recurrent_states = torch.zeros(hidden_states[:, 0].shape, dtype=acc_dtype, device=hidden_states.device)
+
+            contextualized_states = torch.zeros_like(hidden_states)
+            for t in range(hidden_states.shape[1]):
+                recurrent_states = recurrent_gate[:, t].type(acc_dtype) * recurrent_states + hidden_states[:, t].type(
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:389]: it would be nice to split this up instead of the line split on `type(...)`

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 16]
```diff
@@ -0,0 +1,938 @@
...
+                )
+                contextualized_states[:, t] = recurrent_states.type(hidden_states.dtype)
+
+        return contextualized_states, recurrent_states
+
+
+class RecurrentGemmaRecurrentBlock(nn.Module):
+    """Griffin and Hawk's recurrent block."""
+
+    conv1d_state = None
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:400]: Same q here

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 17]
```diff
@@ -0,0 +1,938 @@
...
+    RECURRENTGEMMA_START_DOCSTRING,
+)
+class RecurrentGemmaPreTrainedModel(PreTrainedModel):
+    config_class = RecurrentGemmaConfig
+    base_model_prefix = "model"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["RecurrentGemmaDecoderLayer"]
+    _skip_keys_device_placement = ["cache"]
+    _supports_flash_attn_2 = False
+    _supports_sdpa = False  # we can't compare with eager for now
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:540]: This is a bit funny as it does support SDPA. What's the compare check happening that means we have to force this to False?

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 18]
```diff
@@ -0,0 +1,938 @@
...
+            )
+            use_cache = False
+
+        if inputs_embeds is None:
+            inputs_embeds = self.embed_tokens(input_ids)
+
+        hidden_states = inputs_embeds
+
+        if use_cache and inputs_embeds.shape[1] != 1:  # TODO let's maybe only call in the `generate`?
+            self._setup_cache(self.config, hidden_states.shape[0], hidden_states.device, hidden_states.dtype)
```
Comment (amyeroberts) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:708]: If possible - it would be cleaner

**Review by amyeroberts — COMMENTED** [review]
Very nice 🔥  Thanks for implementing! 

Only thing that has to be addressed is the config saving and some outstanding todo comments dotted around

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py (hunk 17 above)
Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:540]: Becuase the test set it against eager

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py (hunk 15 above)
Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:389]: agreed!

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 19]
```diff
@@ -0,0 +1,938 @@
...
+            hidden_states=normalized_x,
+            recurrent_gate=recurrent_gate,
+            reset=reset,
+            recurrent_states=self.recurrent_states,
+        )
+        self.recurrent_states = recurrent_states
+        return hidden_states
+
+    # TODO refactor
+    def _rnn_scan(
```
Comment (proger) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:349]: Hi! I work on https://github.com/proger/accelerated-scan which is a package with CUDA/Triton kernels. It can speed up training and prompt reading for Recurrent Gemma. Would you consider a patch for the model that uses Accelerated Scan?

Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:349]: Not sure but would love to share if you have a repo on the hub that integrates this? 🤗
